src/common/meta/src/kv_backend/postgres.rs

```diff
@@ -0,0 +1,834 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use snafu::ResultExt;
+use tokio_postgres::{Client, NoTls};
+
+use super::{KvBackend, TxnService};
+use crate::error::{ConnectPostgresSnafu, Error, FromUtf8Snafu, PostgresFailedSnafu, Result};
+use crate::kv_backend::txn::{Txn as KvTxn, TxnResponse as KvTxnResponse};
+use crate::kv_backend::KvBackendRef;
+use crate::rpc::store::{
+    BatchDeleteRequest, BatchDeleteResponse, BatchGetRequest, BatchGetResponse, BatchPutRequest,
+    BatchPutResponse, CompareAndPutRequest, CompareAndPutResponse, DeleteRangeRequest,
+    DeleteRangeResponse, PutRequest, PutResponse, RangeRequest, RangeResponse,
+};
+use crate::rpc::KeyValue;
+
+pub struct PgStore {
+    client: Client,
+}
+
+const METADKV_CREATION: &str =
+    "CREATE TABLE IF NOT EXISTS metakv(k varchar PRIMARY KEY, v varchar);";
+
+impl PgStore {
+    pub async fn with_url(url: &str) -> Result<KvBackendRef> {
+        let (client, conn) = tokio_postgres::connect(url, NoTls)
+            .await
+            .context(ConnectPostgresSnafu)?;
+        tokio::spawn(async move { conn.await.context(ConnectPostgresSnafu) });
+        Self::with_pg_client(client).await
+    }
+
+    pub async fn with_pg_client(client: Client) -> Result<KvBackendRef> {
+        // This step ensures the postgres metadata backend is ready to use.
+        // We check if metakv table exists, and we will create a new table
+        // if it does not exist.
+        client
+            .execute(METADKV_CREATION, &[])
+            .await
+            .context(PostgresFailedSnafu)?;
+        Ok(Arc::new(Self { client }))
+    }
+}
+
+/*
+SELECT k, v FROM metakv
+WHERE k >= start_key and k <= end_key
+*/
+fn generate_range_scan_query(req: RangeRequest) -> Result<String> {
+    let mut sql = String::new();
+    sql.push_str("SELECT k");
+    if !req.keys_only {
+        sql.push_str(",v");
+    }
+    sql.push_str(" FROM metakv");
+    if req.key != b"\0" || req.range_end != b"\0" {
+        let start_key = process_bytes(req.key.clone(), "GetRequestKey")?;
+        match req.range_end.is_empty() {
+            true => sql.push_str(&format!(" WHERE k = '{start_key}'")),
+            false => match is_prefix_range(req.key.clone(), &req.range_end) {
+                true => {
+                    sql.push_str(&format!(" WHERE k LIKE '{start_key}%'"));
+                }
+                false => {
+                    sql.push_str(&format!(" WHERE k >= '{start_key}'"));
+                    if req.range_end != b"\0" {
+                        let end_key = process_bytes(req.range_end, "GetRequestRangeEnd")?;
+                        sql.push_str(&format!(" AND k < '{end_key}'"));
+                    }
+                }
+            },
+        }
+    }
+    // This is a hack that we fetch limit + 1 and compare the result rows against request's limit
```

Comment:
This is because range scan response has a more field -  i want to avoid another cte block such as `(select count(1) from metakv where <predicates> limit limit + 1`
```
pub struct RangeResponse {
    pub kvs: Vec<KeyValue>,
    pub more: bool,
}
```